### PR TITLE
Fix Microsoft.AzureCLI 2.53.1 x64/x86 InstallerUrls

### PR DESCRIPTION
**Pull request by halter73**
This now matches what we did for [2.53.0](https://github.com/microsoft/winget-pkgs/blob/5389630e67e59dedf3162be815c18d61b9976586/manifests/m/Microsoft/AzureCLI/2.53.0/Microsoft.AzureCLI.installer.yaml).

I noticed this when I tried to `winget upgrade --all` and afterwards running `winget upgrade` still showed Microsoft.AzureCLI as out of date at 2.53.0. In turned out I had the 2.53.0 x64 and 2.53.1 x86 installed side by side.

This exact PR has been opened by community members for at least the last three versions: 

- 2.51.0: #116957 @fsdrw08 @BrandonWanHuanSheng
- 2.52.0: #119228 @kkamegawa
- 2.53.0: #121214 @aavdberg @Seros 

@stephengillie @johnstep You probably want to look into fixing the @azclibot. Is that open source?

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-pkgs/pull/123641)

### CODE REVIEW

File: manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml
Comment (mdanish-kh) [inline manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml:10]:
```suggestion
  InstallerSha256: 57b8e2e4bf02059865eab97ba717ae576d9c45621be85b85045e4b1ed51c6efc
  ProductCode: '{8912DDBF-70DA-42F0-939A-278E4B42577D}'
```

File: manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml
Comment (mdanish-kh) [inline manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml:14]:
```suggestion
  InstallerSha256: 50996b7b802e88fa6f5f7eb31b7713b275d4bcbc9676b83c04800ff5e308f4a6
  ProductCode: '{0A56C26A-2E46-4B19-94EB-8482AD658D53}'
```

**Review by mdanish-kh — CHANGES_REQUESTED** [review]

**Comment by mdanish-kh** [comment]
> @stephengillie @johnstep You probably want to look into fixing the @azclibot. Is that open source?

One more thing if the maintainer of the bot sees this PR:

- The PR should also include adding ProductCodes in the installer metadata (it helps with package correlation for MSIs). Upgrade to the latest version of winget-create (if that's what being used in its automation) so that ProductCodes are added/updated automatically in new PRs

**Comment by halter73** [comment]
> You probably want to look into fixing the @azclibot. Is that open source?

/cc @yonzhan @jiasli @bebound

**Comment by bebound** [comment]
I'm not familiar with winget, can we add this info after installation?
```
Winget installs the 64-bit CLI on 64-bit OS by default now. If you used the 32-bit CLI before, please follow this guide to migrate to 64-bit version: https://review.learn.microsoft.com/cli/azure/install-azure-cli-windows?branch=live&tabs=azure-cli#migrate-to-64-bit-azure-cli
```

File: manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml
Comment (bebound) [reply manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml:10]: @mdanish-kh How do you get `ProductCode`?

Comment (mdanish-kh) [reply manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml:10]: @bebound If you use one of the popular [manifest authoring tools](https://github.com/microsoft/winget-pkgs#authoring-a-manifest) to create the manifest, they will automatically add/update this field for you.

If you want to retrieve this for yourself, you can use a number of different methods:

- For installed MSI packages, you can get it by running the command (On PowerShell 5 / Windows PowerShell)

```
get-wmiobject Win32_Product | Sort-Object Name | Format-Table IdentifyingNumber, Name -AutoSize
```

- If you have the `.msi` file, you can use something like [MSI Viewer](https://apps.microsoft.com/detail/msi-viewer/9MZTR9QS01GN) or some PowerShell scripting to retrieve the ProductCode.

- You can run the [manifest in Windows Sandbox using the sandbox script](https://github.com/microsoft/winget-pkgs#in-windows-sandbox) and it will print out the related metadata for the package which includes the ProductCode

**Comment by halter73** [comment]
Is anyone going to approve or merge this? It leads to issues like what's described in https://github.com/Azure/azure-cli/issues/27492#issuecomment-1778497275 as long as anyone tries to upgrade the package. Every hour we wait, more and more people are likely running into problems.

I don't know enough about the packaging rules to know if we should be adding `ProductCode`s if the @azclibot automation doesn't account for it going forward, but I personally don't care either way. We should definitely fix the bot before this happens for a fifth time though.

**Comment by mdanish-kh** [comment]
@halter73 I'll approve this for merge after https://github.com/microsoft/winget-pkgs/pull/123641#discussion_r1370815230 is resolved. ProductCode is one of the fields WinGet client uses for package correlation and it's better to include it in the manifest for MSI packages.

**Comment by mdanish-kh** [comment]
[Policy] reset feedback
[Policy] Needs-Author-Feedback

**Review by mdanish-kh — APPROVED** [review]

File: manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml
Comment (jiasli) [reply manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml:10]: I am able to get the `ProductCode` with the first method on my Windows machine (even though `Get-WmiObject Win32_Product` is slow):

```
> Get-WmiObject Win32_Product | Sort-Object Name | Format-Table IdentifyingNumber, Name, Version -AutoSize

IdentifyingNumber                      Name                                                           Version
-----------------                      ----                                                           -------
...
{C56C2883-5C38-4789-B6BA-4D4B17440AE6} Microsoft Azure CLI (64-bit)                                   2.54.0
```

All these methods require either installing the MSI or other dependencies. As @azclibot is currently hosted on a Linux agent, these methods may not be practicable unless we move @azclibot to a Windows agent.

I am not sure if `ProductCode` comes from `Product.Id`:

https://github.com/Azure/azure-cli/blob/aeb94ed12a46acdd14e7c4e2633908e149f4a1bc/build_scripts/windows/Product.wxs#L36

```
  <Product Id="*"
```

If so, is it possible to have a fixed GUID, instead of generating a new one for each MSI package? cc @heaths

**Comment by jiasli** [comment]
@mdanish-kh, I have created issue https://github.com/Azure/azure-cli/issues/27861 to track adding `ProductCode` to our manifest.

> some PowerShell scripting to retrieve the ProductCode.

It seems all other methods require either installing the MSI or other dependencies. Could you please share more detail on this PowerShell scripting? I believe this script will be very helpful for all winget publishers.

I am also wondering if it is possible to programmatically acquire `ProductCode` during the build process.

File: manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml
Comment (jiasli) [reply manifests/m/Microsoft/AzureCLI/2.53.1/Microsoft.AzureCLI.installer.yaml:10]: As this conversation has been closed, I commented directly on the PR: https://github.com/microsoft/winget-pkgs/pull/123641#issuecomment-1814363236

**Comment by jiasli** [comment]
Thank you @mdanish-kh, but the `Revision Number` provided by `file` command and Windows Properties GUI doesn't seem to match the result of `Get-WmiObject Win32_Product`. For example, for https://azcliprod.blob.core.windows.net/msi/azure-cli-2.54.0-x64.msi, the `Revision Number` is `{8F41B266-5A95-4922-9144-C8BBE1BC8345}`:

```
$ curl https://azcliprod.blob.core.windows.net/msi/azure-cli-2.54.0-x64.msi -O
$ file azure-cli-2.54.0-x64.msi
azure-cli-2.54.0-x64.msi: Composite Document File V2 Document, Little Endian, Os: Windows, Version 10.0, MSI Installer, Code page: 1252, Title: Installation Database, Subject: Microsoft Azure CLI (64-bit), Author: Microsoft Corporation, Keywords: Installer, Comments: This installer database contains the logic and data required to install Microsoft Azure CLI (64-bit)., Template: x64;1033, Revision Number: {8F41B266-5A95-4922-9144-C8BBE1BC8345}, Create Time/Date: Wed Nov  8 05:26:20 2023, Last Saved Time/Date: Wed Nov  8 05:26:20 2023, Number of Pages: 200, Number of Words: 2, Name of Creating Application: Windows Installer XML Toolset (3.10.4.4718), Security: 2
```

image

But after installing the MSI, `Get-WmiObject Win32_Product` shows `IdentifyingNumber` as `{C56C2883-5C38-4789-B6BA-4D4B17440AE6}`:

```
> Get-WmiObject Win32_Product | Sort-Object Name | Format-Table IdentifyingNumber, Name, Version -AutoSize
...
{C56C2883-5C38-4789-B6BA-4D4B17440AE6} Microsoft Azure CLI (64-bit)                                   2.54.0
```

Is there anything I am doing wrong?

**Comment by jiasli** [comment]
BTW, it seems the script of `Get-MSIProperty` can't handle invalid file path. Even if I remove the `try...catch...` statement, the error is pretty vague:

```pwsh
> $MSIFilePath = "$env:USERPROFILE\Downloads\azure-cli-invalid.msi"
> Get-MSIProperty -MSIPath $MSIFilePath -Parameter 'ProductCode'
MethodInvocationException:
Line |
  11 |      $database = $windowsInstaller.GetType().InvokeMember('OpenDatabas …
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Exception calling "InvokeMember" with "5" argument(s): "OpenDatabase,DatabasePath,OpenMode"
```

**Comment by jiasli** [comment]
Our MSI expert @heaths suggests using the `MSI` PowerShell package: https://www.powershellgallery.com/packages/MSI/3.3.4

```pwsh
> Install-Module MSI -Force
> (Get-MSIProperty -Property ProductCode -Path "$env:USERPROFILE\Downloads\azure-cli-2.54.0-x64.msi").Value
{C56C2883-5C38-4789-B6BA-4D4B17440AE6}
```

This is so far the easiest and most direct way I am aware of to retrieve `ProductCode` of an MSI.

**Comment by mdanish-kh** [comment]
@jiasli - it looks like I didn't test the script thoroughly enough for the Linux part. Thanks for finding out the correct way, I should probably look to update the YamlCreate.ps1 script too with this finding

**Comment by mdanish-kh** [comment]
@jiasli Did you get it to work in a Linux environment? I could get the module to work on Windows but still no luck on Linux 👀 


image

**Comment by jiasli** [comment]
@mdanish-kh, you are welcome. I didn't try it on Linux either. 

As described in https://github.com/Azure/azure-cli/issues/27861, I think I am going to extract `ProductCode` while building the MSI on a Windows agent, save the `ProductCode` to an ADO artifact as the metadata of the MSI, then use it to build the manifest on a Linux agent.

However, I am not an expert on MSI or the [`MSI` PowerShell module](https://www.powershellgallery.com/packages/MSI/3.3.4). I think you may report this issue / feature request to the source repo of `MSI` PowerShell module: https://github.com/heaths/psmsi

**Comment by heaths** [comment]
It's not going to work on a non-Windows platform because it uses Windows Installer APIs that are only on Windows. Can you not do this on a Windows agent? Otherwise, there is a Rust crate I've used in another project that actually implements the binary format (OLE docs + proprietary compression algorithms MSIs use). I could write a simple CLI that, given a path, extracts `ProductCode` or any other property if you *really* need this to work on a linux agent.

But the `MSI` module will never be cross-platform. That is **WAY** too much effort for no good reason because you can't use/install MSIs on other platforms, and much of the functionality is geared toward install/modify/uninstall and querying machine state.

**Comment by heaths** [comment]
Oh, look at that: I basically already wrote it: https://github.com/heaths/msigetprop-rs

image

The only difference is that this requires downloading the repo and installing Rust. You can download an x86_64 version here: https://github.com/heaths/msigetprop-rs/releases/download/v0.1.0/msigetprop-x86_64-unknown-linux-gnu.tar.gz

If you need another build, let me know or feel free to install Rust and build whatever you need yourself.

**Pull request merged**
